src/Utilities/Compiler/SmallDictionary.cs

```diff
@@ -68,6 +68,114 @@ private int GetHashCode(K k)
             return Comparer.GetHashCode(k);
         }
 
+        public bool IsEmpty => _root == null;
+
+        public void Remove(K key)
```

Comment:
Thank you for adding this functionality @buyaa-n! I am going to skip reviewing this helper in detail as it seems to be the standard AVL node removal algorithm. Hopefully in future, such a SmallDictionary data structure can be added to BCL and we will not require to maintain this type in both Roslyn repo and this repo.

Reply:
Yes, it is by the standard algorithm, manually tested with several scenarios plus new tests added with this PR also testing it. 

> Hopefully in future, such a SmallDictionary data structure can be added to BCL and we will not require to maintain this type in both Roslyn repo and this repo

Yes, I remember cc @terrajobst was pointing that out before